File: c/2015/day01a/day01a.c
#include <stdio.h>
#include <stdlib.h>

void usage(char *progname) {
    fprintf(stderr, "usage: %s <input file>\n", progname);
    exit(1);
}

int process(char *filename) {
    FILE *infile;
    char ch;
    int floor = 0;
  
    if ( (infile = fopen(filename, "r")) == NULL ) {
        perror("failed to open input file");
        exit(1);
    }
    while ( !feof(infile) ) {
        ch = (char)fgetc(infile);
        switch ( ch ) {
            case '(':
                floor++;
                break;
            case ')':
                floor--;
                break;
        }
    }
    fclose(infile);
    return floor;
}

int main(int argc, char *argv[]) {
    char *filename;
    int result;

    if ( argc < 2 ) {
        usage(argv[0]);
    }

    filename = argv[1];
    result = process(filename);
    printf("result = %d\n", result);
    return 0;
}

File: pascal/2015/day01a/day01a.pas
PROGRAM Day01a(Input, Output);

VAR
    FileName : PChar;
    Res      : Integer;

PROCEDURE Usage();
VAR
    ProgName : PChar;
BEGIN
    ProgName := argv[0];
    WriteLn('usage: ', ProgName, ' <input file>');
    Halt(1)
END;

FUNCTION Process(FileName : PChar) : Integer;
VAR
    InFile  : Text;
    Ch      : Char;
    Floor   : Integer;

BEGIN
    Floor := 0;
    {$I-}
    Assign(InFile, FileName);
    Reset(InFile);
    IF IOResult <> 0 THEN
    BEGIN
        WriteLn(StdErr, 'error when opening file: ', FileName);
        Halt(1)
    END;
    {$I+}
    WHILE NOT EOF(InFile) DO
    BEGIN
        Read(InFile, Ch);
        CASE Ch OF
            '(' : Floor := Floor + 1;
            ')' : Floor := Floor - 1;
        END;
    END;
    Close(InFile);
    Process := Floor
END;

BEGIN
    IF argc < 2 THEN
        Usage();
    FileName := argv[1];
    Res := Process(FileName);
    WriteLn('result = ', Res)
END.
